controllers/clustergroupupgrade_controller.go

```diff
@@ -1561,15 +1628,43 @@ func (r *ClusterGroupUpgradeReconciler) getAllClustersForUpgrade(ctx context.Con
 	return clusterNames, nil
 }
 
+func (r *ClusterGroupUpgradeReconciler) updateChildResourceNames(ctx context.Context, safeNameMap map[string]string, childResourceNameList []string, u *unstructured.Unstructured) ([]string, error) {
```

Comment:
I think we need a description for this function. Also, it's unclear what "childResource" means just from looking at the function. We're also not updating names in this function, we're returning a list of names. Maybe something like "GetUpdatedResourceNameList"?

I'm confused about a couple of things here:
1. Why are we deleting resources here... Is it if it happens to create multiple policies/placement rules/placement bindings for the same CGU, but with different names?
2. Given how this function is called, how do you know you are deleting the oldest created object? Does it even matter what object gets deleted?
3. It was my understanding that the ResourceNameAnnotation is keeping the desired name, but at line 1634 we are comparing it to the safe name.

Maybe it will all make more sense tomorrow morning :smiley:

Reply:
Yes, I can add a description. For your questions:
1. Yes, whenever we see multiple items from the query with the same desired name in their annotation, we need to correct the situation by deleting the extra, as they won't get updated if the managed policy content changes. Only the one being tracked by the name map should be kept.
2. This kind of conflict only happens when the reconcile call starts with a stale status/name map (empty or missing names) and creates new names (i.e. new resources) in the map. At the end, before updating status, this function is used like a sanity check. The ones that are already in the map just get created in this reconcile call.
3. No, u.GetName() is the safe name of the object. I guess I will change the variable name to desiredName.